src/main/java/org/spongepowered/common/data/persistence/DataViewJsonWriter.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.persistence;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.io.CharStreams;
+import com.google.gson.stream.JsonWriter;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataQuery;
+import org.spongepowered.api.data.DataSerializable;
+import org.spongepowered.api.data.DataView;
+import org.spongepowered.api.data.MemoryDataContainer;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link JsonWriter} that serializes to a {@link DataContainer}
+ * instead of a JSON string.
+ *
+ * Inspired by GSON's {@link com.google.gson.internal.bind.JsonTreeWriter}.
+ */
+public final class DataViewJsonWriter extends JsonWriter {
+
+    private static final Writer UNWRITABLE_WRITER = new Writer() {
+        @Override public void write(char[] buffer, int offset, int counter) {
+            throw new AssertionError();
+        }
+        @Override public void flush() throws IOException {
+            throw new AssertionError();
+        }
+        @Override public void close() throws IOException {
+            throw new AssertionError();
+        }
+    };
+
+    private final List<Object> stack = new ArrayList<>();
```

Comment:
I use the same as Gson itself is using. I don't really see a reason for using an actual `Stack`, the `ArrayList` will do the job just fine :P